source/slang/slang-ir-spirv-legalize.cpp

```diff
@@ -372,6 +373,123 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
         addUsersToWorkList(ptrType);
     }
 
+    void processLoop(IRLoop* loop)
+    {
+
+        // 2.11.1. Rules for Structured Control-flow Declarations
+        // Structured control flow declarations must satisfy the following
+        // rules:
+        //   - the merge block declared by a header block must not be a merge
+        //     block declared by any other header block
+        //   - each header block must strictly structurally dominate its merge
+        //     block
+        //   - all back edges must branch to a loop header, with each loop
+        //     header having exactly one back edge branching to it
+        //   - for a given loop header, its merge block, OpLoopMerge Continue
+        //     Target, and corresponding back-edge block:
+        //       - the Continue Target and merge block must be different blocks
+        //       - the loop header must structurally dominate the Continue
+        //         Target
+        //       - the Continue Target must structurally dominate the back-edge
+        //         block
+        //       - the back-edge block must structurally post dominate the
+        //         Continue Target
+
+        // If the continue block has only a single predecessor, pretend like it
+        // is just ordinary control flow
+        //
+        // TODO: could this fail in cases like this, where it had a single
+        // predecessor, but it's still nested inside a region?
+        // do{
+        //   if(x)
+        //     continue;
+        //   unreachable
+        // } while(foo)
+        const auto t = loop->getTargetBlock();
+        auto c = loop->getContinueBlock();
+        if(c->getPredecessors().getCount() <= 1)
+        {
+            c = t;
+            loop->continueBlock.set(c);
+        }
+
+        // Our IR allows multiple back-edges to a loop header if this is also
+        // the loop continue block. SPIR-V does not so replace them with a
+        // single intermediate block
+        if(c == t)
+        {
+            // Subtract one predecessor for the loop entry
+            const auto numBackEdges = c->getPredecessors().getCount() - 1;
+
+            // If we have multiple back-edges, make a new block at the end of
```

Comment:
If we do this, could the jump to the newBlock violate structured rules (e.g. if the original continue is in some nested if statement, then the new jump could be treated like a normal jump but crosses the boundary of its parent merge points).
Does SPIRV allow that?

Reply:
It seems that the safest way to avoid violating those restrictions is to just call eliminateContinue and then elimianteMultiLevelBreak.

Reply:
> jump to the newBlock violate structured rules

I think this transformation is safe, at the very least it's no less safe than it was before.